src/transaction/log_recovery.c

```diff
@@ -994,94 +1051,2176 @@ log_recovery_finish_transactions (THREAD_ENTRY * const thread_p)
+		     (long long int) log_Gl.hdr.append_lsa.pageid, log_Gl.hdr.append_lsa.offset);
+    }
+
+  return;
+}
+
+static int
+log_rv_analysis_check_page_corruption (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, const LOG_PAGE * log_page_p,
+				       corruption_checker & checker)
+{
+#if !defined(NDEBUG)
+  if (prm_get_bool_value (PRM_ID_LOGPB_LOGGING_DEBUG))
+    {
+      _er_log_debug (ARG_FILE_LINE, "logpb_recovery_analysis: log page %lld, checksum %d\n",
+		     log_page_p->hdr.logical_pageid, log_page_p->hdr.checksum);
+      fileio_page_hexa_dump ((const char *) log_page_p, LOG_PAGESIZE);
+    }
+#endif /* !NDEBUG */
+
+  checker.check_page_checksum (log_page_p);
+  if (checker.is_page_corrupted ())
+    {
+      if (logpb_is_page_in_archive (pageid))
+	{
+	  /* Should not happen. */
+	  logpb_fatal_error (thread_p, true, ARG_FILE_LINE, "log_rv_analysis_check_page_corruption");
+	  return ER_FAILED;
+	}
+
+      checker.find_first_corrupted_block (log_page_p);
+
+      /* Found corrupted log page. */
+      if (prm_get_bool_value (PRM_ID_LOGPB_LOGGING_DEBUG))
+	{
+	  _er_log_debug (ARG_FILE_LINE, "logpb_recovery_analysis: log page %lld is corrupted due to partial flush"
+			 " (first_corrupted_rec_lsa = %lld|%d)\n",
+			 (long long int) pageid, LSA_AS_ARGS (&checker.get_first_corrupted_lsa ()));
+	}
+    }
+  return NO_ERROR;
+}
+
+static void
+log_rv_analysis_handle_fetch_page_fail (THREAD_ENTRY * thread_p, log_recovery_context & context, LOG_PAGE * log_page_p,
+					const LOG_RECORD_HEADER * log_rec, const log_lsa & prev_lsa,
+					const log_lsa & prev_prev_lsa)
+{
+  if (context.is_restore_from_backup ())
+    {
+      context.set_forced_restore_stop ();
+
+#if !defined(NDEBUG)
+      if (prm_get_bool_value (PRM_ID_LOG_TRACE_DEBUG))
+	{
+	  fprintf (stdout, msgcat_message (MSGCAT_CATALOG_CUBRID, MSGCAT_SET_LOG, MSGCAT_LOG_STARTS));
+	  fprintf (stdout,
+		   msgcat_message (MSGCAT_CATALOG_CUBRID, MSGCAT_SET_LOG,
+				   MSGCAT_LOG_INCOMPLTE_MEDIA_RECOVERY), context.get_end_redo_lsa ().pageid,
+		   context.get_end_redo_lsa ().offset, "???...\n");
+	  fprintf (stdout, msgcat_message (MSGCAT_CATALOG_CUBRID, MSGCAT_SET_LOG, MSGCAT_LOG_STARTS));
+	  fflush (stdout);
+	}
+#endif /* !NDEBUG */
+
+      /* if previous log record exists, reset tdes->tail_lsa/undo_nxlsa as previous of end_redo_lsa */
+      if (log_rec != NULL)
+	{
+	  LOG_TDES *last_log_tdes = LOG_FIND_TDES (logtb_find_tran_index (thread_p, log_rec->trid));
+	  if (last_log_tdes != NULL)
+	    {
+	      LSA_COPY (&last_log_tdes->tail_lsa, &log_rec->prev_tranlsa);
+	      LSA_COPY (&last_log_tdes->undo_nxlsa, &log_rec->prev_tranlsa);
+	      if (prm_get_bool_value (PRM_ID_LOGPB_LOGGING_DEBUG))
+		{
+		  _er_log_debug (ARG_FILE_LINE, "logpb_recovery_analysis: trid = %d, tail_lsa=%lld|%d\n",
+				 log_rec->trid, last_log_tdes->tail_lsa.pageid, last_log_tdes->tail_lsa.offset);
+		}
+	    }
+	}
+
+      assert (!prev_lsa.is_null ());
```

Comment:
How about removing this assertion? I think `prev_lsa` can be null in the case below.

1. `prev_lsa` is set to NULL_LSA at the beginning of `log_recovery_analysis_internal ()` ( log_recovery.c:2539L)
2. `logpb_fetch_page()` can fail in `log_recovery_analysis_internal ()` (log_recovery.c:2591L)
3. `log_rv_analysis_handle_fetch_page_fail ()` can be called with `prev_lsa`(NULL_LSA).

or How about throwing it as an exception? 

if it requires much modification, I think it is better to do it in other issue.

Reply:
Thrown an exception - fatal. Thanks.